### PR TITLE
Factor ProductMaps.hpp into a tpp file

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Factor the ProductMaps function definitions into a `tpp` file so that we don't have to parse this every time we include anything related to the maps.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/Elliptic/Actions/Test_InitializeSystem.cpp [hunk 1]
```diff
@@ -16,6 +16,7 @@
 #include "Domain/CoordinateMaps/Affine.hpp"
 #include "Domain/CoordinateMaps/CoordinateMap.hpp"
 #include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.tpp"
```
Comment (moxcodes) [inline tests/Unit/Elliptic/Actions/Test_InitializeSystem.cpp:19]: Is this one necessary? The only place the `ProductOf...Maps` appears is in `PUPable_reg`, and I don't know whether that pattern constructs the object.

File: tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeDomain.cpp [hunk 2]
```diff
@@ -12,6 +12,7 @@
 #include "Domain/CoordinateMaps/Affine.hpp"
 #include "Domain/CoordinateMaps/CoordinateMap.hpp"
 #include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.tpp"
```
Comment (moxcodes) [inline tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeDomain.cpp]: Same question: only used in `PUPable_reg`

File: tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeInterfaces.cpp [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #include "Domain/CoordinateMaps/Affine.hpp"
 #include "Domain/CoordinateMaps/CoordinateMap.hpp"
 #include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.tpp"
```
Comment (moxcodes) [inline tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeInterfaces.cpp:17]: Same

File: tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp [hunk 4]
```diff
@@ -14,6 +14,7 @@
 #include "Domain/CoordinateMaps/Affine.hpp"
 #include "Domain/CoordinateMaps/CoordinateMap.hpp"
 #include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.tpp"
```
Comment (moxcodes) [inline tests/Unit/ParallelAlgorithms/DiscontinuousGalerkin/Test_InitializeMortars.cpp:17]: And this one too.

**Review by moxcodes — DISMISSED** [review]
Looks good. There's no change in functionality, but there were a few of the .tpp includes where the only use that I could find in the file were in `PUPable_reg`, and it's unclear to me whether that requires the implementation file.

For all of the files in question, though, it probably doesn't matter much because the `tpp` is indirectly included from the domain creators.

**Comment by nilsdeppe** [comment]
@moxcodes you're right, not needed. Pushed a fixup

**Comment by moxcodes** [comment]
looks good, I'm happy for you to squash.

**Comment by nilsdeppe** [comment]
@moxcodes squashed! Thanks for the quick review!

**Review by moxcodes — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@moxcodes I had to add a couple includes back to get GCC to link the code. I squashed directly and rebased on develop.

**Comment by kidder** [comment]
@gsb76 please look at

**Review by moxcodes — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
Rebased on develop and included `ProductMaps.hpp` as an include in the `tpp`

**Review by moxcodes — APPROVED** [review]

**Review by mar-celine — APPROVED** [review]
Should check that these changes are consistent with how other tpp files are included in cpp files (if the hpp file is included or not), otherwise looks good!

**Comment by nilsdeppe** [comment]
Looking through the code base there seems to be a mix of whether `hpp`s are included with `tpp`s. @wthrowe @kidder do you have a preference here? We still don't have that many `tpp` files and bringing in a new rule would be easy at this point.

**Review by kidder — APPROVED** [review]

**Pull request merged**